common/http-api-client/src/dns.rs

```diff
@@ -0,0 +1,153 @@
+        self
+    }
+}
+
+struct SocketAddrs {
+    iter: LookupIpIntoIter,
+}
+
+#[derive(Debug)]
+struct HickoryDnsSystemConfError(ResolveError);
+
+/// Wrapper around an `AsyncResolver`, which implements the `Resolve` trait.
+#[derive(Debug, Default, Clone)]
+pub struct HickoryDnsResolver {
+    /// Since we might not have been called in the context of a
+    /// Tokio Runtime in initialization, so we must delay the actual
+    /// construction of the resolver.
+    state: Arc<OnceCell<TokioAsyncResolver>>,
+    fallback: Arc<OnceCell<TokioAsyncResolver>>,
+}
+
+impl Resolve for HickoryDnsResolver {
+    fn resolve(&self, name: Name) -> Resolving {
+        let resolver = self.state.clone();
+        let fallback = self.fallback.clone();
+        Box::pin(async move {
+            let resolver = resolver.get_or_try_init(new_resolver)?;
+
+            // try the primary DNS resolver that we set up (DoH or DoT or whatever)
+            let lookup = match resolver.lookup_ip(name.as_str()).await {
+                Ok(res) => res,
+                Err(e) => {
+                    // on failure use the fall back system configured DNS resolver
+                    warn!("primary DNS failed w/ error {e}: using system fallback");
+                    let resolver = fallback.get_or_try_init(new_resolver_system)?;
+                    resolver.lookup_ip(name.as_str()).await?
+                }
+            };
+
+            let addrs: Addrs = Box::new(SocketAddrs {
+                iter: lookup.into_iter(),
+            });
+            Ok(addrs)
+        })
+    }
+}
+
+impl Iterator for SocketAddrs {
+    type Item = SocketAddr;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|ip_addr| SocketAddr::new(ip_addr, 0))
+    }
+}
+
+/// Create a new resolver with a custom DoT based configuration. The options are overridden to look
+/// up for both IPv4 and IPv6 addresses to work with "happy eyeballs" algorithm.
+fn new_resolver() -> Result<TokioAsyncResolver, HickoryDnsSystemConfError> {
+    let mut name_servers = NameServerConfigGroup::google_tls();
+    name_servers.merge(NameServerConfigGroup::google_https());
+    // name_servers.merge(NameServerConfigGroup::google_h3());
+    name_servers.merge(NameServerConfigGroup::quad9_tls());
+    name_servers.merge(NameServerConfigGroup::quad9_https());
+    name_servers.merge(NameServerConfigGroup::cloudflare_tls());
+    name_servers.merge(NameServerConfigGroup::cloudflare_https());
+
+    let config = ResolverConfig::from_parts(None, Vec::new(), name_servers);
+
+    let mut opts = ResolverOpts::default();
+    opts.ip_strategy = LookupIpStrategy::Ipv4thenIpv6;
+    // Would like to enable this when 0.25 stabilizes
+    // opts.server_ordering_strategy = ServerOrderingStrategy::RoundRobin;
+
+    Ok(TokioAsyncResolver::tokio(config, opts))
+}
+
+/// Create a new resolver with the default configuration, which reads from `/etc/resolve.conf`. The
```

Comment:
you're saying it reads from `/etc/resolve.conf`. obviously this file doesn't exist on all operating systems, so how does it work on, for example, windows?

Reply:
This is a function of the `hickory_resolver` library. Their implementation is [generic across unix and windows](https://docs.rs/hickory-resolver/latest/src/hickory_resolver/system_conf/mod.rs.html). The `/etc/resolv.conf` is there to say "for example". This comment, and a solid portion of the code for this resolver, came from [the resolver that reqwest implements using hickory](https://docs.rs/reqwest/latest/src/reqwest/dns/hickory.rs.html). The reqwest resolver that uses hickory is not externally configurable in the way I needed so I extended it.  I will update the comment too be more generic for platform target though so as not to be misleading.